pkg/informer/informer.go

```diff
@@ -0,0 +1,179 @@
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+const (
+	resyncPeriod   = 10 * time.Hour
+	pollTypesEvery = time.Minute
+)
+
+type DynamicDiscoverySharedInformerFactory struct {
+	disco   *discovery.DiscoveryClient
+	dsif    dynamicinformer.DynamicSharedInformerFactory
+	gvrs    sets.String
+	handler GVREventHandler
+}
+
+// IndexerFor returns the indexer for the given type GVR.
+func (d *DynamicDiscoverySharedInformerFactory) IndexerFor(gvr schema.GroupVersionResource) cache.Indexer {
+	return d.dsif.ForResource(gvr).Informer().GetIndexer()
+}
+
+// Listers returns a slice of per-resource-type listers for all types that are
+// known by this informer factory.
+func (d *DynamicDiscoverySharedInformerFactory) Listers() []cache.GenericLister {
+	var listers []cache.GenericLister
+	for _, gvrstr := range d.gvrs.UnsortedList() {
+		gvr, _ := schema.ParseResourceArg(gvrstr)
+		if gvr == nil {
+			klog.Errorf("parsing GVR string %q", gvrstr)
+			continue
+		}
+
+		listers = append(listers, d.dsif.ForResource(*gvr).Lister())
+	}
+	return listers
+}
+
+// NewDynamicDiscoverySharedInformerFactory returns an informer that discovers
+// new types and informs on updates to resources of those types.
+func NewDynamicDiscoverySharedInformerFactory(
+	disco *discovery.DiscoveryClient,
+	dynClient dynamic.Interface,
+	handler GVREventHandler) DynamicDiscoverySharedInformerFactory {
+	dsif := dynamicinformer.NewDynamicSharedInformerFactory(dynClient, resyncPeriod)
+	return DynamicDiscoverySharedInformerFactory{
+		disco:   disco,
+		dsif:    dsif,
+		gvrs:    sets.NewString(),
+		handler: handler,
+	}
+}
+
+// GVREventHandler is an event handler that includes the GroupVersionResource
+// of the resource being handled.
+type GVREventHandler interface {
+	OnAdd(gvr schema.GroupVersionResource, obj interface{})
+	OnUpdate(gvr schema.GroupVersionResource, oldObj, newObj interface{})
+	OnDelete(gvr schema.GroupVersionResource, obj interface{})
+}
+
+type GVREventHandlerFuncs struct {
+	AddFunc    func(gvr schema.GroupVersionResource, obj interface{})
+	UpdateFunc func(gvr schema.GroupVersionResource, oldObj, newObj interface{})
+	DeleteFunc func(gvr schema.GroupVersionResource, obj interface{})
+}
+
+func (g GVREventHandlerFuncs) OnAdd(gvr schema.GroupVersionResource, obj interface{}) {
+	if g.AddFunc != nil {
+		g.AddFunc(gvr, obj)
+	}
+}
+func (g GVREventHandlerFuncs) OnUpdate(gvr schema.GroupVersionResource, oldObj, newObj interface{}) {
+	if g.UpdateFunc != nil {
+		g.UpdateFunc(gvr, oldObj, newObj)
+	}
+}
+func (g GVREventHandlerFuncs) OnDelete(gvr schema.GroupVersionResource, obj interface{}) {
+	if g.DeleteFunc != nil {
+		g.DeleteFunc(gvr, obj)
+	}
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) Start(ctx context.Context) {
+	// Immediately discover types and start informing.
+	if err := d.discoverTypes(ctx); err != nil {
+		klog.Errorf("Error discovering initial types: %v", err)
```

Comment:
Should we return early here, if discovery fails the very first time?

Reply:
I've changed it to a `Fatal`, but I don't love it. We could retry in case it's flaky, but that's effectively what logging and starting the ticker did before.